src/modsim2/similarity/metrics/kde.py

```diff
@@ -0,0 +1,295 @@
+        s = AB_distances[:, -1]
+
+        # Calculate distance according to corrected eqn 14 in paper
+        distance = np.log(s / r).sum() * d / n + np.log(m / (n - 1))
+
+        return distance
+
+    def calculate_distance(
+        self,
+        data_A: np.ndarray,
+        data_B: np.ndarray,
+        labels_A: np.ndarray,
+        labels_B: np.ndarray,
+        embedding_name: str,
+        kernel_name: str,
+        metric_name: str,
+        embedding_kwargs: dict = {},
+        integration_kwargs: dict = {},
+    ) -> float:
+        """
+        Calculates the distance between datasets A and B using kernel
+        density estimation to represent the probability density function
+        of the datasets.
+
+        The L2 norm and total variation are options that are provided
+        to calculate the distance between the distributions. The KL
+        divergence between the distributions can also be calculated;
+        however, this is not a true metric as it is not symmetrical.
+        An approximate method for calculating the KL divergence is
+        provided, although this does not require the density estimation.
+
+        Args:
+            data_A: The first image dataset
+            data_B: The second image dataset
+            labels_A: The target values for the first dataset
+            labels_B: The target values for the second dataset
+            embedding_name: The feature embeddings to use for the input arrays
+            kernel_name: The kernel to be applied in the density estimation, any
+                    kernels allowed in sklearns KernelDensity class are valid.
+                    Currently these are 'gaussian', 'tophat', 'epanechnikov',
+                    'exponential', 'linear', 'cosine'
+            metric_name: The metric to be used to calculate the distance between
+                    the density functions, currently only 'l2' is valid
+            embedding_kwargs: A dictionary of key word arguments to be passed
+                    to the embedding function
+            integration_kwargs: A dictionary of key word arguments to be passed
+                    to the integration function
+        """
+        # A value of zero is returned if the datasets are the same (the integration
+        # methods cannot be used if the datasets are the same, and the approximate
+        # kl method will calculate as -inf)
+        if np.array_equal(data_A, data_B):
+            return 0
+
+        # Check for valid embedding choice
+        assert embedding_name in EMBEDDING_FN_DICT, "Error: embedding does not exist"
+
+        if embedding_name in ["inception_umap"]:
+            # as UMAP is stochastic, a random seed is required
+            # for reproducability
+            embed_random_seed = randint(low=0, high=100, size=(1,))[0]
+            embedding_kwargs["random_seed"] = int(embed_random_seed)
+
+        # Embed the data
+        embed_A, embed_B = self._embed_data(
```

Comment:
may have misunderstood but shouldn't this embedding happen in `_pre_process_data`?